Update WebRTC media source stats (#32538)
mdn/content: files/en-us/web/api/rtcvideosourcestats/framespersecond/index.md

---
title: "RTCVideoSourceStats: framesPerSecond property"
short-title: framesPerSecond
slug: Web/API/RTCVideoSourceStats/framesPerSecond
page-type: web-api-instance-property
browser-compat: api.RTCStatsReport.type_media-source.framesPerSecond
---

{{APIRef("WebRTC")}}

> The **`framesPerSecond`** property of the {{domxref("RTCVideoSourceStats")}} dictionary indicates the number of frames from this video source in the last second.

Author: A `RTCVideoSourceStats` represents stats for a video media track attached to a sender. The spec says "originating from". I suspect "sent" and "generated by"  are effectively synonyms in this case. But how about we use the spec words, so we can't be wrong.

```suggestion
The **`framesPerSecond`** property of the {{domxref("RTCVideoSourceStats")}} dictionary indicates the number of frames originating from this video source in the last second.
```

The property is not defined on the stats object for the first second of its lifetime.

## Value

> A number indicating the frames from this source in the last second.

Author: ```suggestion
A number indicating the frames originating from this source in the last second.
```

## Examples

This example shows how you might iterate the stats object returned from `RTCRtpSender.getStats()` to get the video source stats, and then extract the `framesPerSecond`.

```js
// where sender is an RTCRtpSender
const stats = await sender.getStats();
let videoSourceStats = null;

stats.forEach((report) => {
  if (report.type === "media-source" && report.kind==="video") {
    videoSourceStats = report;
    break;
  }
});

// Note, test is conditional because

> // framesPerSecond does not exist in the first second

Reviewer: This isn't accurate, `?.` here protects you against `videoSourceStats` being undefined or null, but not `framesPerSecond`: that is, it applies to the left hand side. It's still a good thing to do here because `videoSourceStats` might be null, and this avoids an error in that situation.

Author: @wbamberg Are you sure? I mean you're right that it protects the LHS (`videoSourceStats`) and not `framesPerSecond` but I don't think it matters. 
- If `videoSourceStats` does not exist then the result short circuits to undefined (`framesPerSecond` also doesn't exist of course).
- If `videoSourceStats` does exist but `framesPerSecond` does not then the result is `undefined` - which is what we want right?

Reviewer: Yes, I agree with all that, I'm just saying the comment is wrong:

> // Note, test is conditional because
> // framesPerSecond does not exist in the first second

The test is conditional because `videoSourceStats` might be `null`. So for instance if the code did this:

```js
// where sender is an RTCRtpSender
const stats = await sender.getStats();
let videoSourceStats = null;

stats.forEach((report) => {
  if (report.type === "media-source" && report.kind==="video") {
    videoSourceStats = report;
   const fps = videoSourceStats.framesPerSecond;
    break;
  }
});
```

...then you wouldn't need `?.` because you know `videoSourceStats` is not null. Of course `framesPerSecond` can still be undefined, but that doesn't matter, as you say, and it's not affected by `?.`.

At least, such is my understanding.

Author: That makes sense. Thanks for clarifying and providing some words.

```suggestion
// Note, test is conditional in case the stats object
// does not include video source stats
```
